blog/content/post/2024-11-26-measured-boot-improvements.md

```diff
@@ -0,0 +1,222 @@
+but with a slight difference - the event log alone is just a chunk of memory
+that the payload can read and parse however they like. Still, each payload
+would have to do the same: interpret coreboot's custom event log to create a
+single and complete event log from the reset vector to the OS. The amount of
+code duplication it would create is unthinkable. Thankfully, we have already
+implemented the TCG standardized event logging, which helps a bit with the
+portability of the parsing.
+
+In this article, I will show how we achieved the complete integrity of the TPM
+event log across the whole boot process of open-source firmware based on
+coreboot and TianoCore EDKII UEFI Payload in Dasharo.
+
+### Solution
+
+We already have many pieces prepared earlier, like unified TPM support and a
+TCG-compliant event log. The work required to pass the vent log from coreboot
+is not that huge, as EDKII already knows the TCG event log format, so all we
+have to do is process each event and create a HandOffBlock (HOB) containing
+the measurements made by stages preceding UEFI DXE (Driver Execution
+Environment). Using HOBs is a standard mechanism of passing information from
+the pre-DXE environment to DXE, where the drivers responsible for publishing
+the event log to the OS reside. It is done in [this Pull
+Request](https://github.com/Dasharo/edk2/pull/139).
+
+Of course, this is not the end of modifications. Depending on the payload,
+some glue may also be required on the coreboot side. We have to teach coreboot
+not to publish TPM ACPI tables (TCPA and TPM2) because they contain the
+address of the TPM event log in the memory. This task is now left to the
+payload, which publishes the respective TPM ACPI table with its own, combined
+TPM event log. It is achieved in the [previously linked Pull Request on EDKII
+side](https://github.com/Dasharo/edk2/pull/139) and also on [coreboot
+side](https://github.com/Dasharo/coreboot/pull/517) (some of the changes here
+are already upstream). Besides the TPM ACPI table, coreboot creates an event
+log using only one hashing algorithm:
+
+- SHA1 - default for TPM1.2
+- SHA256 - default for TPM2.0
+
+TPM2.0 log format supports agile measurements, i.e. there may be multiple
+measurements of the same component using multiple hashing algorithms. For
+simplicity, coreboot uses only one currently. The choice of the hashing
+algorithm must also be passed to the UEFI Payload. It is done by [specifying a
+mask of supported hashing algorithms at UEFI Payload build
+time](https://github.com/Dasharo/coreboot/pull/517/files#diff-58bb01776a00af5a5d61de00d6ea3008bcd67177fb7991fb41e8976c6b4d046a).
+
+Now that all the pieces are in place, we can enjoy having a consistent and
+complete TPM event log in the Dasharo, effectively solving the [issue of not
+reproducible PCR2
+value](https://github.com/Dasharo/dasharo-issues/issues/455).
+
+## Summary
+
+TPM event log integrity is an excellent achievement for our Dasharo team. While we
```

Comment:
```suggestion
TPM event log integrity is an excellent achievement for our Dasharo Team. While we
```

Reply:
https://github.com/3mdeb/news-and-ideas/pull/613/commits/cff1006b43a13b63a542f10eed3a227911bce453#diff-053de20f23ef9f4af660d4f062e10dc313dc72f5ede66ec1a080c9c82ed815e6R294